File: rails/app/models/import_history.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: import_histories
#
#  id            :bigint           not null, primary key
#  user_id       :bigint           not null
#  unit_id       :bigint
#  file_name     :string(255)      not null
#  file_size     :bigint
#  content_type  :string(255)
#  status        :integer          default("pending"), not null
#  total_count   :integer          default(0), not null
#  success_count :integer          default(0), not null
#  error_count   :integer          default(0), not null
#  started_at    :datetime
#  finished_at   :datetime
#  deleted_at    :datetime
#  created_at    :datetime         not null
#  updated_at    :datetime         not null
#  mode          :integer          default("append"), not null
#  import_type   :integer          default("question"), not null
#
class ImportHistory < ApplicationRecord
  belongs_to :user
  belongs_to :unit, optional: true
  has_many :import_errors, -> { order(:row_number) }, dependent: :destroy, inverse_of: :import_history

  has_one_attached :file

  validates :unit, presence: true, if: :question?

  scope :active, -> { where(deleted_at: nil) }

  enum status: {
    pending: 0,
    processing: 1,
    completed: 2,
    failed: 3
  }

  enum :mode, {
    append: 0,
    overwrite: 1
  }, validate: true

  enum :import_type, {
    question: 0,
    student: 1
  }, validate: true
end

File: frontend/types/common/import_history.ts
export type ImportStatus = "pending" | "processing" | "completed" | "failed";

export type ImportMode = "append" | "overwrite";

export type ImportHistory = {
  id: number;
  file_name: string;
  status: ImportStatus;
  success_count: number;
  error_count: number;
  total_count: number;
  created_at: string;
};

// インポート履歴のフィルタ・サマリーで参照する関連レコードの最小表現。
// 一覧のプルダウンと詳細のサマリーで同じ形を使う。
export type CourseOption = { id: number; level_name: string };
export type UnitOption = { id: number; unit_name: string };
export type UserOption = { id: number; name: string };
